src/mawpy/steps/update_stay_duration.py

```diff
@@ -0,0 +1,80 @@
+import logging
+
+import psutil
+import time
+import pandas as pd
+
+from multiprocessing import Pool
+from multiprocessing import cpu_count
+
+from mawpy.constants import UNIX_START_T, USER_ID, STAY_DUR, STAY_LAT, STAY_LONG, STAY_UNC, STAY
+from mawpy.utilities.preprocessing import get_preprocessed_dataframe, get_list_of_chunks_by_column
+
+logger = logging.getLogger(__name__)
+
+
+def _get_stay_duration_for_group(df_per_group: pd.DataFrame) -> pd.DataFrame:
+    """
+        For the traces with the same STAY, computes the stay duration by calculating the difference between
+        timestamps of last and first trace
+    """
+    df_per_group[STAY_DUR] = (df_per_group.iloc[-1][UNIX_START_T] +
+                              max(0, df_per_group.iloc[-1][STAY_DUR]) -
+                              df_per_group.iloc[0][UNIX_START_T])
+    return df_per_group
+
+
+def _run_for_user(df_by_user: pd.DataFrame, duration_constraint: float, order_of_execution: int = 1) -> pd.DataFrame:
+    """
+        Calculate duration of stays for each user.
+    """
+    if order_of_execution == 1:
+        df_by_user[STAY_DUR] = -1
+
+    df_by_user = df_by_user.groupby(STAY).apply(lambda x: _get_stay_duration_for_group(x))
+    df_by_user.loc[df_by_user[STAY_LAT] == -1, STAY_DUR] = -1
+    df_by_user.loc[df_by_user[STAY_DUR] < duration_constraint, [STAY_LAT, STAY_LONG, STAY_UNC, STAY_DUR]] = (
+        -1, -1, -1, -1)
+
+    return df_by_user
+
+
+def _run(args: tuple) -> pd.DataFrame:
+    df_by_user, dur_constraint = args
+    df_by_user = _run_for_user(df_by_user, dur_constraint)
+    return df_by_user
+
+
+def update_stay_duration(output_file: str, dur_constraint: float, input_df: pd.DataFrame | None, input_file: str = None) -> pd.DataFrame:
```

Comment:
This looks almost identical to `def trace_segmentation_clustering` could this be a reusable function?

Reply:
Thanks! Have extracted the function for parallel execution across incremental_clustering, update_stay_duration and trace_segmentation_clustering

Reply:
Very nice